### PR TITLE
Audit use buffered backend

### LINKED ISSUE #53020: [audit] Restore audit logging in the scalability tests
As a result of https://github.com/kubernetes/kubernetes/issues/51899, audit logging was disabled in the scalability tests by https://github.com/kubernetes/test-infra/pull/4720

After https://github.com/kubernetes/kubernetes/issues/53006 is fixed, those tests should again have audit logging enabled to verify the scalability of this feature.

/cc @tallclair

### PR DESCRIPTION
**Pull request by crassirostris**
This is the next step after https://github.com/kubernetes/kubernetes/pull/60076

This PR fixes https://github.com/kubernetes/kubernetes/issues/53020, to address https://github.com/kubernetes/kubernetes/issues/53006 later

In this PR buffered backend, introduced in https://github.com/kubernetes/kubernetes/pull/60076, is used to replace ad-hoc solution for webhook and add an ability to enable buffering for the log audit backend.

```release-note
Log audit backend can now be configured to perform batching before writing events to disk.
```

/cc @sttts @tallclair @ericchiang @CaoShuFeng

### CODE REVIEW

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	// events are still coming after the backend was stopped.
+	var sendErr error
+	var evIndex int
+
+	// If the delegateBackend was shutdown and the buffer channel was closed, an
+	// attempt to add an event to it will result in panic that we should
+	// recover from.
+	defer func() {
+		if err := recover(); err != nil {
+			sendErr = fmt.Errorf("panic when processing events: %v", err)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go]: nit: I liked the old error message more. The new one sounds like something is broken.

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+// Shutdown blocks until stopCh passed to the Run method is closed and all
+// events added prior to that moment are batched and sent to the delegate backend.
+func (b *bufferedBackend) Shutdown() {
+	// Wait until the routine spawned in Run method exits.
+	<-b.shutdownCh
+
+	// Wait until all sending routines exit.
+	b.wg.Wait()
```
[the remark this answers is not in the document]
Comment (crassirostris) [reply staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go]: Sounds good, added a comment

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go (hunk 1 above)
Comment (crassirostris) [reply staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go]: Agree, done

**Review by crassirostris — COMMENTED** [review]
@sttts Thanks for the review, comments addressed

**Comment by sttts** [comment]
/approve
Lgtm, but would like to have some more eyes one it, especially a detailed review of the changes to the old backends. @tallclair are you available?

**Comment by crassirostris** [comment]
/assign @thockin 

Could you please approve this PR for the new dependency?

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 3]
```diff
@@ -264,39 +351,20 @@ func (o *AuditWebhookOptions) AddFlags(fs *pflag.FlagSet) {
...
-	fs.Float32Var(&o.BatchConfig.ThrottleQPS, "audit-webhook-batch-throttle-qps",
-		o.BatchConfig.ThrottleQPS, "Maximum average number of requests per second. "+
-			"Only used in batch mode.")
-	fs.IntVar(&o.BatchConfig.ThrottleBurst, "audit-webhook-batch-throttle-burst",
-		o.BatchConfig.ThrottleBurst, "Maximum number of requests sent at the same "+
-			"moment if ThrottleQPS was not utilized before. Only used in batch mode.")
-	fs.DurationVar(&o.BatchConfig.InitialBackoff, "audit-webhook-batch-initial-backoff",
-		o.BatchConfig.InitialBackoff, "The amount of time to wait before retrying the "+
-			"first failed requests. Only used in batch mode.")
+	fs.DurationVar(&o.InitialBackoff, "audit-webhook-initial-backoff",
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: Unfortunately you can't just change flags (see [deprecation policy, #5b](https://kubernetes.io/docs/reference/deprecation-policy/#deprecating-a-flag-or-cli)). I suggest marking the old one as deprecated, and making it an alias for the new one.

**Review by tallclair — COMMENTED** [review]
lgtm, except the flag issue

**Review by crassirostris — COMMENTED** [review]
@tallclair Thanks! PTAL

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 4]
```diff
@@ -264,39 +351,23 @@ func (o *AuditWebhookOptions) AddFlags(fs *pflag.FlagSet) {
...
+		o.InitialBackoff, "The amount of time to wait before retrying the "+
+			"first failed requests.")
+	fs.DurationVar(&o.InitialBackoff, "audit-webhook-batch-initial-backoff",
+		o.InitialBackoff, "The amount of time to wait before retrying the first failed "+
+			"requests. Deprecated, use --audit-webhook-initial-backoff instead.")
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: Use [`fs.MarkDeprecated(...)`](https://godoc.org/github.com/spf13/pflag#FlagSet.MarkDeprecated)

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: Thanks! Was deceived by [an example in another file](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go#L131)

Done

**Comment by liggitt** [comment]
audit e2e tests started flaking as soon as this merged. tracking in https://github.com/kubernetes/kubernetes/issues/60719

File: staging/src/k8s.io/apiserver/pkg/server/options/audit.go [hunk 5]
```diff
@@ -67,27 +96,37 @@ type AuditLogOptions struct {
...
+			BatchOptions: AuditBatchOptions{
+				Mode:        ModeBatch,
+				BatchConfig: defaultLogBatchConfig,
+			},
+			InitialBackoff: pluginwebhook.DefaultInitialBackoff,
+		},
+		LogOptions: AuditLogOptions{
+			Format: pluginlog.FormatJson,
+			BatchOptions: AuditBatchOptions{
+				Mode:        ModeBatch,
```
[the remark this answers is not in the document]
Comment (CaoShuFeng) [reply staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: There is already a pull request: https://github.com/kubernetes/kubernetes/pull/60739

[the remark this answers is not in the document]
Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/server/options/audit.go]: I know, this is a conscious decision. I believe batching is always better by default than blocking

**Pull request merged**